### DISCUSSION TITLE
Lesson 1: Are computationally expensive pure functions free of gas cost?

### DISCUSSION THREAD

**Discussion by shashibheemanapally** [smartcontractkit/full-blockchain-solidity-course-py#680] [Q&A] [answered]
Hi, I have two questions.
1. In the video, it was stated that `pure` functions are free of gas cost. But assume I have a following pure function which is computationally expensive.
```Solidity

function getAnswer(uint256 _a,uint256 _b) public pure returns(uint256){
    uint256  ans;
    for(int i=0;i<1000000000;i++){
      //do somtehing ans=f(_a,_b)
    }
    return ans;
}
```
Is the above function still free to call? If yes, how is the node which is doing all the computation getting rewarded?
And what happens if the function results in some kind of stackoverflow?

2.Do you think blockchain is economically viable than centralized server-client architecture ? (not sure if I am missing something!). I mean, for a very simple transaction i.e changing the value of a variable costing me 0.0001eth(~0.4$). Of course blockchain gives some good characteristics over the other. Your thoughts?

Thank you

**Comment by cromewar** [comment, marked as answer]
Hello @shashibheemanapally 
1. This is tricky as is not likely anyone would do something like this on a function ever, but let's suppose it happens yo have to separate two different things, nodes just get paid when they have record anything new on the blockchain (new transactions, new contracts, etc) but as pure functions does not need something to be recorded the gas price will be always free. In the other hand the one responsible to run this pure functions is the EVM (Ethereum virtual machine) a whole operating system inside the blockchain, which has it's own memory ram and etc, if you run something like this, the EVM is most likely going to run for a long period of time until it get's the result. As the resources the EVM has are very limited is not probable it would represent a real problem for the nodes storing it, if you get a stackoverflow the transaccion will get reverted 
2. Well, depends a lot of what you want to accomplish because not everything should be stored on the blockchain, there are some scenarios than you will prefer just use a traditional server-client architecture. You would like to store something that's important and it must last forever on the blockchain, like economic transactions. You also can use both, like Moralis which provides you MongoDB based servers for those info which should not be stored on the blockchain.

Also, you are right this is actually expensive because we are using Ethereum network, but there are other blockchain specialized on other things which are a lot cheaper, and anyways this should be solved when ethereum 2.0 arrives.

**Reply by shashibheemanapally** [reply]
Hi @cromewar , in point 1 , you are saying that the pure functions are free...! Don't you think people will take the advantage of all these resources of EVM for free? Like for example some numerical algorithms, machine learning algorithms etc., only need numerical inputs. I can deploy smart contracts including these algorithms and I can use these for free of cost!!

**Reply by cromewar** [reply]
Not really, The EVM has it's limitations to prevent something like this, this is really low level stuff, the EVM has 140 unique instructional code called `Opcodes` which are limited to 1 byte of computational power, making the top of it 256. Each Opcode comes with it's own gas cost. Sure, pure functions has no gas cost as they are just for math purposes, but they are ruled by the max capacity of the EVM if your calculations are that heavy your transaction will revert getting a RollOver error (happens when you exceed the maximum capacity returning it to 0).

However I have to admit that what you are saying is theoretically possible, like create infinite loops which refresh the memory time to time, but I don't see any scenario where this could be useful to anyone at all.
